tests/integration/test_norm.py

```diff
@@ -76,16 +94,63 @@ def test_axis_1d(ndim, ord, keepdims):
 @pytest.mark.parametrize("ndim", range(2, LEGATE_MAX_DIM + 1))
 @pytest.mark.parametrize("ord", MATRIX_ORDS)
 @pytest.mark.parametrize("keepdims", [False, True])
-def test_axis_2d(ndim, ord, keepdims):
+@pytest.mark.parametrize(
+    "axis",
+    ((0, 1), pytest.param((1, 0), marks=pytest.mark.xfail)),
+    ids=lambda axis: f"(axis={axis})",
+)
+def test_axis_2d(ndim, ord, keepdims, axis):
+    # For all cases when axis is (1, 0) and ord is None or fro,
+    # output values of cuNumeric and Numpy are different and not close enough
     np_res = np.linalg.norm(
-        np_arrays[ndim], ord=ord, axis=(0, 1), keepdims=keepdims
+        np_arrays[ndim], ord=ord, axis=axis, keepdims=keepdims
     )
     num_res = num.linalg.norm(
-        num_arrays[ndim], ord=ord, axis=(0, 1), keepdims=keepdims
+        num_arrays[ndim], ord=ord, axis=axis, keepdims=keepdims
     )
     assert allclose(np_res, num_res)
 
 
+class TestNormErrors:
+    def test_axis_invalid_type(self):
+        # In cuNumeric, raises error in normalize_axis_tuple
+        x = num.array([1, 2, 3])
+        axis = "string"
+        with pytest.raises(TypeError):
+            num.linalg.norm(x, axis=axis)
+
+    @pytest.mark.parametrize(
+        "axis",
+        (3, -4, (1, 1), (1, 3), (1, 0, 2)),
+        ids=lambda axis: f"(axis={axis})",
+    )
+    def test_axis_invalid_value(self, axis):
+        # for (1, 1), In cuNumeric, raises error in normalize_axis_tuple
+        ndim = 2
+        with pytest.raises(ValueError):
+            num.linalg.norm(num_arrays[ndim], axis=axis)
```

Comment:
@robinw0928  I feel like these exception type comparisons could be improved by having the tests  run the code in both numpy and cunumeric and actually verifying that exceptions do, in fact, match in both cases. i.e. I think it would be possible for these tests to all look something like:
```python
    @pytest.mark.parametrize("axis", (3, -4, (1, 1), (1, 3), (1, 0, 2)))
    def test_axis_invalid_value(self, axis, exc_compare):
        exc_compare("linalg.norm", ValueError, num_arrays[ndim], axis=axis)
```

The `exc_compare` would look maybe like (untested sketch):
```python
def exc_compare(func_name, exc, *args, **kw):
    try: 
        # run code with numpy
        # np.func(*args, **kw)
    except exc as e: 
        exc__np = exc 
    except Exception as e:
        exc_np = type(e)
    else: 
        exc_np = None

    try: 
        # run code with cunumeric
        # num.func(*args, **kw)
    except exc as e: 
        exc_num = exc 
    except Exception as e:
        exc_num = type(e)
    else: 
        exc_num = None

    assert exc_np == exc
    assert exc_num == exc
```

There'd be some details to sort out, specifically how to turn string func names like `"linalg.norm"` in to actual functions `np.linalg.norm` and `num.linalg.norm`  but a helper for that would actually be nice to have and useful in other contexts.

What do you think?

Reply:
@bryevdv 
I tried your approach and found that we have to call both np and num in test code as the input array is different for np and num.

> @pytest.mark.parametrize("axis", (3, -4, (1, 1), (1, 3), (1, 0, 2)))
def test_axis_invalid_value(self, axis, exc_compare):
        exc_compare("np", "linalg.norm", ValueError, **np_arrays**[ndim], axis=axis)
        exc_compare("num", "linalg.norm", ValueError, **num_arrays**[ndim], axis=axis)

If this is the case, I think we don't need to change and directly call 

> with pytest.raises(ValueError):
>  num.linalg.norm(num_arrays[ndim], axis=axis)

What do you think?

Reply:
@robinw0928 My interest is to try to make sure, *via test*, that both libraries generate the *same* exception. Currently we maintain that the cunumeric codepaths generate some given exception, but it's never actually checked that that exception matches numpy. All we have are some comments about numpy, and comments are hard to keep up to date and maintain. Also if we do an actual comparison we could easily mark specific cases as `xfail` right in the parametrize list (just as an example): 

```python
@pytest.mark.parametrize("axis", (
    3, 
    # cunumeric result differs for negative axis
    pytest.param(-4, marks=pytest.mark.xfail),
    (1, 1), 
    (1, 3), 
    (1, 0, 2)))
def test_axis_invalid_value(self, axis):
```
So, I might be misunderstanding your suggestion, but can you walk me through how it would maintain, under actual test, that the exceptions from both libraries match?

Reply:
@bryevdv 
I understand what you mean. 
Currently, in negative tests, we only call cunumeric. So exception from numpy is not checked. If we need to maintain the exceptions from both libraries match, the code is like:

> @pytest.mark.parametrize("axis", (3, -4, (1, 1), (1, 3), (1, 0, 2)))
> def test_axis_invalid_value(self, axis):
>         ndim = 2
>         with pytest.raises(ValueError):
>             np.linalg.norm(**np_arrays**[ndim], axis=axis)
>         with pytest.raises(ValueError):
>             num.linalg.norm(**num_arrays**[ndim], axis=axis)

Do you have any suggestion?

Reply:
@robinw0928 FYI you can get proper code quoting if you use the `<>` toolbar icon, or triple backtick ``` code fences around the code block

> Do you have any suggestion?

Having to write out both manually is somewhat tedious, but maybe it is necessary. I do think, by way of a convention, it would be slightly more robust to factor out the exception type in a variable, just to make sure that both branches use the same type:
```python
expected_exc = ValueError

with pytest.raises(expected_exc):
    np.linalg.norm(np_arrays[ndim], axis=axis)

with pytest.raises(expected_exc):
    num.linalg.norm(num_arrays[ndim], axis=axis)
```

That way there is less chance the two branches accidentally change to disagree.

Reply:
Fixed. 
Thanks for your FYI. You are so kind.
```
expected_exc = ValueError

with pytest.raises(expected_exc):
    np.linalg.norm(np_arrays[ndim], axis=axis)

with pytest.raises(expected_exc):
    num.linalg.norm(num_arrays[ndim], axis=axis)
```
triple backtick ``` really works.